guardian/cdk
# feat: add first draft of EC2 App Pattern

## What does this change?
This PR adds a first attempt at our EC2 App pattern. It's still a work in progress, but we have confirmed that this minimal version of the pattern can be deployed successfully so it feels like a good time to get something merged.

## Does this change require changes to existing projects or CDK CLI?
No.

## How to test
We've added a basic unit test for this. More unit tests (and proper docs) will be added as the pattern develops.

We've also manually tested deploying a stack with the following code:

```typescript
import type { App } from "@aws-cdk/core";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import { GuEc2App } from "@guardian/cdk/lib/patterns/ec2-app";

export class GuCDKEC2Pattern extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const port = 8000;

    new GuEc2App(this, {
      app: "TestApp",
      applicationPort: port,
      publicFacing: false,
      userData: `#!/usr/bin/env bash
        mkdir test
        cd test
        echo OK >> healthcheck
        python -m SimpleHTTPServer ${port}
      `,
    });
  }
}
```

This just allows us to serve a static file via `/healthcheck`, so that we can ensure that requests can be served correctly.

## How can we measure success?
It should be much faster/easier to create a new EC2 app from scratch using this pattern.

## Have we considered potential risks?
Although this pattern is now functional, we don't want 'real' users to start experimenting with it at this stage. To mitigate against this:

1) We've added a [comment](https://github.com/guardian/cdk/pull/399/commits/ab1f0eb408a8653c945ab8ddfd9d20d7044776a7) to warn people that this is still under development
2) The pattern is not mentioned in https://github.com/guardian/cdk/blob/main/src/patterns/index.ts yet.
